torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -69,21 +72,37 @@ ProcessGroupAgent::ProcessGroupAgent(
       threadPool_(numSendRecvThreads) {
   TORCH_CHECK(nameMap_.size() > 1, "ProcessGroupAgent requires world_size to "
       "be at least 2, but got ", nameMap_.size());
-  auto workerRankIter = nameMap_.find(workerName_);
+  auto workerRankIter = nameMap_.find(workerId_.name_);
   TORCH_CHECK(workerRankIter != nameMap_.end(), "Failed to resolve worker "
-      "name ", workerName_, " to a ProcessGroup rank.");
+      "name ", workerId_.name_, " to a ProcessGroup rank.");
   TORCH_CHECK(pg_->getRank() == workerRankIter -> second,
       "Resolved worker rank ", workerRankIter -> second,
       " does not match ProcessGroup rank ", pg_->getRank());
 
-  names_.resize(nameMap_.size());
-  for (auto& entry : nameMap_) {
-    names_[entry.second] = entry.first;
+  // tmp vector to sort names in rank's order
+  std::vector<std::string> tmpWorkerIds(pg_->getSize());
+  for (auto& entry: nameMap_) {
+    tmpWorkerIds[entry.second] = entry.first;
   }
+
+  workerIds_.reserve(pg_->getSize());
+  for (int rank = 0; rank < (int)tmpWorkerIds.size(); ++rank) {
+    workerIds_.emplace_back(std::move(tmpWorkerIds[rank]), rank);
+  }
+
   PythonRpcHandler::init();
   listenerThread_ = std::thread(&ProcessGroupAgent::listenLoop, this);
 }
 
+const WorkerId& ProcessGroupAgent::getWorkerId(
+    const std::string& workerName) const {
+  const auto idIter = nameMap_.find(workerName);
+  TORCH_CHECK(idIter != nameMap_.end(),
```

Comment:
Would be good to add a few unit tests that trigger errors like these, ex:

1. Pass an invalid worker name.
2. Pass invalid worker id.
3. Pass worker name as self.
4. Pass worker id as self.

Reply:
Yes, will add those, thanks!